### PR TITLE
issue #140 - compatibility doctrine/dbal:^4

### PR DESCRIPTION
**Pull request by pounard**

### CODE REVIEW

File: src/Anonymization/Anonymizator.php [hunk 1]
```diff
@@ -434,21 +441,22 @@ protected function createUpdateQuery(string $table): Update
...
-                $builder->select($table),
+                $this->session->select($table),
                 $expr->where()->isEqual(
                     $expr->column(AbstractAnonymizer::JOIN_ID, $table),
                     $expr->column(AbstractAnonymizer::JOIN_ID, AbstractAnonymizer::JOIN_TABLE),
                 ),
                 AbstractAnonymizer::JOIN_TABLE
             );
-        } elseif (Platform::SQLITE === $builder->getServerFlavor()) {
+        } elseif ($this->session->vendorIs(Vendor::SQLITE)) {
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizator.php:451]: I like this :+1:

File: src/Anonymization/Anonymizator.php [hunk 2]
```diff
@@ -597,50 +595,32 @@ protected function removeAnonymizerIdColumn(string $table): void
...
-        );
-
-        $queryBuilder = $this->getQueryBuilder();
+        $this
+            ->session
+            ->getSchemaManager()
+            ->modify()
+            ->addColumn($table, AbstractAnonymizer::JOIN_ID, Type::intBig(), true)
+            ->commit()
+        ;
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizator.php:604]: So much simple 👌

File: src/Anonymization/Anonymizer/AbstractAnonymizer.php [hunk 3]
```diff
@@ -26,7 +22,7 @@ abstract class AbstractAnonymizer
     final public function __construct(
         protected string $tableName,
         protected string $columnName,
-        protected Connection $connection,
+        protected Bridge $session,
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizer/AbstractAnonymizer.php:25]: Same as above, maybe `$dbSession` or `$databaseSession`

File: src/Anonymization/Anonymizer/AbstractAnonymizer.php [hunk 4]
```diff
@@ -173,51 +170,36 @@ protected function createSampleTempTable(array $columns, array $values = [], arr
...
-                        \array_map(fn ($column) => ':'. $column, $columns)
-                    ))
-                    ->setParameters(\array_combine($columns, $value))
-                    ->executeQuery()
-                ;
+        foreach ($values as $key => $value) {
+            // Allow single raw value when there is only one column.
+            $value = (array) $value;
+            if ($columnCount !== \count($value)) {
+                throw new \InvalidArgumentException(\sprintf("Row %s in sample list column count (%d) mismatch with table column count (%d)", $key, \count($values), $columnCount));
```
Comment (SimonMellerin) [inline src/Anonymization/Anonymizer/AbstractAnonymizer.php:196]:
```suggestion
                throw new \InvalidArgumentException(\sprintf(
                    "Row %s in sample list column count (%d) mismatch with table column count (%d)", 
                    $key, 
                    \count($values), 
                    $columnCount
                ));
```

File: src/Test/UnitTestCase.php [hunk 5]
```diff
@@ -31,6 +30,10 @@
 abstract class UnitTestCase extends TestCase
 {
     private bool $initialized = false;
+    /**
+     * @deprecated
+     *   Remove when hard dependency on doctrine/dbal will is gone.
```
Comment (SimonMellerin) [inline src/Test/UnitTestCase.php:35]:
```suggestion
     *   Remove when hard dependency on doctrine/dbal will be gone.
```

File: src/Test/UnitTestCase.php [hunk 6]
```diff
@@ -48,20 +51,15 @@ protected function disconnect(): void
      * Create database connection.
      *
      * Default will use a mock object.
+     *
+     * @deprecated
+     *   Remove when hard dependency on doctrine/dbal will is gone.
```
Comment (SimonMellerin) [inline src/Test/UnitTestCase.php:56]:
```suggestion
     *   Remove when hard dependency on doctrine/dbal will be gone.
```

File: src/Anonymization/Anonymizator.php [hunk 7]
```diff
@@ -33,15 +30,17 @@ class Anonymizator implements LoggerAwareInterface
...
     public function __construct(
-        private Connection $connection,
+        Connection $connection,
         private AnonymizerRegistry $anonymizerRegistry,
         private AnonymizationConfig $anonymizationConfig,
         private ?string $salt = null,
     ) {
         $this->logger = new NullLogger();
         $this->output = new NullOutput();
+        $this->session = new DoctrineQueryBuilder($connection);
```
[the remark this answers is not in the document]
Comment (pounard) [reply src/Anonymization/Anonymizator.php:43]: I'm OK with this change, but is that really necessary ?

[the remark this answers is not in the document]
Comment (SimonMellerin) [reply src/Anonymization/Anonymizator.php:43]: I think `$session` is too generic

**Comment by pounard** [comment]
I really do not understand why the CS fixer errors... When I run it locally everything's fine.

**Review by SimonMellerin — APPROVED** [review]

**Comment by SimonMellerin** [comment]
Thanks !

**Pull request merged**